### ISSUE TITLE
feature request: Implement certbots --dry-run functionality

### ISSUE THREAD

**Issue by ClundXIII** [Neilpang/acme.sh#2565]
After running into issue #2463 I realized that --test cannot be used to replace --dry-run from certbot.

Something equivalently to certbots --dry-run is definitely needed. Especially if you tinker with your webserver configuration and need to test if the well-known directory is still accessible. There needs to be a simple way to test the configuration without having to issue a new certificate and having the old one being overwritten by --test.

Regards,
Simon

**Comment by tobiasmuecksch** [comment]
This feature would be extremely helpful!

**Comment by AvverbioPronome** [comment]
Has any of you both done anything about this?

I am trying but I find this feature pretty hard to implement, because I cannot just write the test cert to `/dev/null` (acme.sh expects to read it to check if it's a fullchain or not, and to install it) and I don't know if using a `mktemp -d` will isolate the configuration enough. (in my experiments, it doesn't interfere with the main acme.sh config, but it will install a a staging certificate into the actual app.)

```
TEMPDIR=$(mktemp -d)

cp -a "$LE_WORKING_DIR/." "$TEMPDIR"

cd "$TEMPDIR"

LE_WORKING_DIR="$TEMPDIR"

./acme.sh --test --renew-all --force\
                       --key-file ""\
                       --fullchain-file ""\
                       --reloadcmd ""

# this still replaces the cert in the app -- do not use.




rm -r "$TEMPDIR"
```

**Comment by ClundXIII** [comment]
nope. I am just using the normal --force function (since I override the old cert). Luckily the old config is preserved if something fails. So it kinda does a dry run if something is wrong.

Also I am hoping I don't hit any ratelimits.

**Comment by kevdogg** [comment]
I'm struggling with this issue myself -- I was surprised that with the --stagging or --test flag, I actually had certs issued from Let's Encrypt.  I don't recall any actual certs being issued when I used to do certbot --dry-run.  

Will I hit any ratelimits?  I'm messing around with a custom script to actually obtain the certs, install the certs, and setup systemd time files all in one shot.  A dry-run would be very helpful

**Comment by AvverbioPronome** [comment]
@kevdogg the staging api has ratelimits, they're just very high. 

Both `acme.sh --test` and `certbot --dry-run` use the staging api, the difference is in what the client does with the certificates it obtains. `certbot` discards them, `acme.sh` deploys them.

**Comment by bits-fritz** [comment]
> Both `acme.sh --test` and `certbot --dry-run` use the staging api, 

For acme.sh this is only true for `--issue` action. 
`--renew` action **does use** the api the certificate was issued with.

> the difference is in what the client does with the certificates it obtains. `certbot` discards them, `acme.sh` deploys them.

No

**Comment by ClundXIII** [comment]
> For acme.sh this is only true for `--issue` action.

I am talking about the `--issue` action. Not sure why someone wants to dry-run renew.

**Comment by neilpang** [comment]
A certbot-style `--dry-run` that discards the obtained certificate will not be added, but the isolation equivalent is simple and fully safe: run the test in a throwaway config home, e.g. `acme.sh --config-home /tmp/acme-dryrun --server letsencrypt_test --issue -d example.com -w /var/www/html`. That validates your webserver / .well-known setup against the staging CA without touching your production certificates or config in any way; delete the directory afterwards. Also note: since 3.1.3, `--renew` strictly honors the CA recorded at issue time and never switches between staging and production (see #2401), so a staging test can no longer leak into production renewals. Closing.
